### PR TITLE
Fix/data race in watchdog

### PR DESCRIPTION
**Pull request by philippseith**
Tried to fix the race in the websocket connections' watchdog. When running go test -race, this one is gone (but the sse race is still there, as #95 hasn't been merged in this branch.
Albeit, I am not sure this watchdog implementation does fulfil its purpose. Specific tests are still missing.

### CODE REVIEW

File: websocketconnection.go [hunk 1]
```diff
@@ -48,67 +51,80 @@ func (w *webSocketConnection) Read(p []byte) (n int, err error) {
...
+		timer:      time.NewTimer(timeout),
+		cancelChan: make(chan struct{}, 1),
+	}
+	var dogCtx context.Context
+	dogCtx, dog.bark = context.WithCancel(ctx)
+	return dogCtx, dog
+}
+
+func (d *watchDog) Cancel() {
+	d.cancelChan <- struct{}{}
```
Comment (andig) [inline websocketconnection.go]: Why not close the channel if new watchdogs are created anyway?

**Comment by andig** [comment]
My feeling (nothing more) is that the entire watchdog is much too complex. It would be nice if we could simplify that down to a channel that either is updated or a websocket connection timer expires. I‘ll try to take a look once this PR has landed.

**Comment by philippseith** [comment]
> My feeling (nothing more) is that the entire watchdog is much too complex. It would be nice if we could simplify that down to a channel that either is updated or a websocket connection timer expires. I‘ll try to take a look once this PR has landed.

The core problem is that you can't simply pass a WithTimeout context to the Read/Write operation, as expiration in one kind of operation will close the websocket, regardless if the other kind of operation was successfully executed.
If you have a server which got a request for streaming (from Read) it calls Read again (waiting for a second request) and starts writing happily. If the client does not send another request, the Read timeout will expire and the websocket will be closed, albeit it was used heavily when writing the stream in the meantime.

**Comment by philippseith** [comment]
I did some experiments with various timeout/keepalive settings in client and server and looked at  the resulting messages  sent between the parties.
Now it seems to me that monitoring the timeout in the connection is the wrong way at all. If the timeout expires, the connection just closes. But the signalr protocol defines a Close message that should be sent when the signalr connection closes that could not be sent when the connection is already closed.
I remember adding a variable Timeout property to the connection because of the handshake process which can have a different timeout value and isn't driven by the normal signalr message loop. 
Adding the timeout handling to the handshake code might complicate it, but this is a one time effort - in contrast to implementing the correct timeout handling for each connection type.
Bonus: The watchdog can be removed.

**Comment by andig** [comment]
Standing by to test ;)

**Comment by philippseith** [comment]
See #99

**Pull request merged**